### PR TITLE
30 dataset importer script

### PR DESCRIPTION
**Pull request by rogup**
#### What? Why?

Related to issue #30

Implement and test a script which can import a CSV of data and write a folder full of the files that the mykomap back-end can use.  

This has been test-integrated with upcoming branches and will integrate with the following exceptions. I needed to:
- Revert "[common] Comment out parts of the prop spec that we don't need in the API"
- Revert "[common] Allow single VocabPropDef instance to provide all the lang translations of a vocab"

Other conflicting changes upstream, specifically the `searchable.json` format had changed unexpectedly (to me), have been adjusted for.

Dependencies have been added for parsing CSV (`udsv`) and a CLI framework (`clipanion`).

#### What should we test?
The code should build and unit tests should all pass as ever.

Usage instructions are included in the documentation added in the command itself, which is defined in `src/cli/dataset.ts`. They are fairly brief for now, as more comprehensive documentation would need to explain the config schema, which is itself a moving target.

You can if you wish to test run the script against the dataset CSV included in the test, or even a larger one, although you will need an accompanying config for that.

A runscript has been added to aid with this:

    npm run -- dataset import --help

This invokes using `tsx`, which is therefore a dependency needed in production mode. `ts-node` was tried originally, but appears to be problematic. I've not found a better way of running this so far.

#### Checklist

- [ ] Updated or added any necessary docs in `docs/`
- [ ] Added UTs
- [ ] Checked that all automated tests pass

### CODE REVIEW

File: apps/back-end/src/dataset.ts [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+  }
+
+  /** A JSON.stringify replacer for objects.
+   *
+   * As above, but only alters  lat/lng properties
+   */
+  limitLatLngDecimals(key: string, value: any): any {
+    if (key === "lat" || key === "lng") {
+      if (value == null) return null; // note: nullish comparison
+      return Number(Number(value).toFixed(6));
```
Comment (rogup) [inline apps/back-end/src/dataset.ts:114]: I think we decided 5 dp's, which equates to around 1m uncertainty in the position of the marker. 6 is probably fine for the size of data we have, but the performance gain will be a lot once we scale up the number of co-ops.
https://open-innovations.org/blog/2023-07-25-tips-for-optimising-geojson-files

**Review by rogup — COMMENTED** [review]
@wu-lee Just a couple of comments but looks good to merge otherwise

File: apps/back-end/test/data/dataset-cli/config.json [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      "from": "Other Activities"
+    },
+    "websites": {
+      "type": "multi",
+      "of": {
+        "type": "value"
+      },
+      "from": "Websites"
+    },
+    "validated": {
```
[the remark this answers is not in the document]
Comment (wu-lee) [reply apps/back-end/test/data/dataset-cli/config.json:89]: Do you mean the "from" attribute?

If so, it's something which was in the original Mykomaps prop-def interfaces which turns out to be handy for configuring the mapping of data from CSV headers in the import script.

If you mean the "validated" thing, that's just a column of booleans which allows that case to be tested.  As in "is this row validated by a human?"

[the remark this answers is not in the document]
Comment (wu-lee) [reply apps/back-end/test/data/dataset-cli/config.json:89]: WRT the needs of the CLI vs the needs of the server/front end (guessing that's what you mean by "server UTs"?), there is arguably a distinction to be made and perhaps the API doesn't need to send them, but the benefits vs cost ratio of separating the two in a maintainable way seems to be low enough that I don't think it's a good thing to try, certainly at this stage.

File: apps/back-end/src/csv.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
+/** CSV related functions */
```
[the remark this answers is not in the document]
Comment (wu-lee) [reply apps/back-end/src/csv.ts:1]: There is a logic to it, which makes sense to me at least. Happy to discuss though.

File: apps/back-end/src/dataset.ts (hunk 1 above)
Comment (wu-lee) [reply apps/back-end/src/dataset.ts:114]: Ok, I don't recall offhand. Happy to change this to 5dp.

File: apps/back-end/src/csv.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rogup) [reply apps/back-end/src/csv.ts:1]: I'm fine with it for now, it was just slightly confusing, will resolve this conversation

**Comment by rogup** [comment]
This work has already made its way into `main` via `nick/dev3` so will close this PR.

**Pull request closed without merging**